File: back/src/main/java/academy/aicode/astrobookings/service/LaunchService.java
package academy.aicode.astrobookings.service;

import academy.aicode.astrobookings.exception.ResourceNotFoundException;
import academy.aicode.astrobookings.model.Launch;
import academy.aicode.astrobookings.model.LaunchStatus;
import academy.aicode.astrobookings.repository.LaunchRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
public class LaunchService {
    private static final Logger LOGGER =
            LoggerFactory.getLogger(LaunchService.class);
    private final LaunchRepository repository;
    private final RocketService rocketService;

    public LaunchService(LaunchRepository repository, RocketService rocketService) {
        this.repository = repository;
        this.rocketService = rocketService;
    }

    public Launch createLaunch(Launch launch) {
        rocketService.getRocketById(launch.getRocketId())
                .orElseThrow(() -> new IllegalArgumentException("Rocket not found with id: " + launch.getRocketId()));

        launch.setId(null);
        launch.setStatus(LaunchStatus.CREATED);
        Launch savedLaunch = repository.save(launch);
        LOGGER.info("Launch created: {} for rocket {} on {}", savedLaunch.getId(), savedLaunch.getRocketId(), savedLaunch.getLaunchDate());
        return savedLaunch;
    }

    public List<Launch> getAllLaunches() {
        return repository.findAll();
    }

    public Launch getLaunchById(UUID id) {
        return repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Launch not found with id: " + id));
    }

    public Launch updateLaunchStatus(UUID id, LaunchStatus status) {
        Launch launch = getLaunchById(id);
        LaunchStatus oldStatus = launch.getStatus();
        launch.setStatus(status);
        Launch updatedLaunch = repository.save(launch);
        LOGGER.info("Launch {} status updated from {} to {}", id, oldStatus, status);
        return updatedLaunch;
    }

}

File: front/src/services/launchService.js
import { apiClient } from './apiClient';

const LAUNCHES_ENDPOINT = '/launches';

export async function getAllLaunches() {
  console.info('Fetching all launches');
  const launchesData = await apiClient(LAUNCHES_ENDPOINT);
  return launchesData;
}

export async function getLaunchById(id) {
  console.info(`Fetching launch with id: ${id}`);
  const launchData = await apiClient(`${LAUNCHES_ENDPOINT}/${id}`);
  return launchData;
}

export async function createLaunch(launch) {
  console.info('Creating new launch', launch);
  const createdLaunch = await apiClient(LAUNCHES_ENDPOINT, { body: launch });
  console.info('Launch created successfully', createdLaunch);
  return createdLaunch;
}

export async function updateLaunchStatus(id, status) {
  console.info(`Updating status for launch ${id} to ${status}`);
  const updatedLaunch = await apiClient(`${LAUNCHES_ENDPOINT}/${id}/status`, {
    method: 'PATCH',
    body: status
  });
  console.info(`Launch status updated successfully for ${id}`, updatedLaunch);
  return updatedLaunch;
}
